Duckling/Types/Document.hs

```diff
@@ -125,9 +125,59 @@ data CharClass
 -- the reasonability of the match to actually be a word.
 isRangeValid :: Lang -> Document -> Int -> Int -> Bool
 isRangeValid = \case
+  AR -> arIsRangeValid
   ZH -> zhIsRangeValid
   _ -> defaultIsRangeValid
   where
+    arIsRangeValid :: Document -> Int -> Int -> Bool
+    arIsRangeValid doc start end =
+      ((start == 0 ||
+            isDifferent (doc ! (start - 1)) (doc ! (start))) &&
+        (end == length doc ||
+            isDifferent (doc ! (end - 1)) (doc ! (end)))) ||
+      -- Is Arabic proclitic?
+      (start == end - 1 &&
+        isArabicProclitic (doc ! start) &&
+        (start == 0 || isDifferent (doc ! (start - 1)) (doc ! start))) ||
+      (start == end - 2 &&
+        isArabicProclitic2 (doc ! start) (doc ! (start + 1)) &&
+        (start == 0 || isDifferent (doc ! (start - 1)) (doc ! start))) ||
+      -- Is preceeded by proclitic
+      (start /= 0 && isArabicProclitic (doc ! (start - 1)) &&
+        (end == length doc ||
+          isDifferent (doc ! (end - 1)) (doc ! (end)))) ||
+      -- Is Arabic enclitic?
+      (start == (end - 2) && isArabicEnclitic (doc ! start) (doc ! (end - 1)) &&
+        (end == length doc || isDifferent (doc ! end) (doc ! (end + 1)))) ||
+      -- Is followed by enclitic
+      ((start ==0 || isDifferent (doc ! (start - 1)) (doc ! (start))) &&
+        (end <= (length doc - 2) &&
+          isArabicEnclitic (doc ! (end)) (doc ! (end + 1))))
+      where
+        -- This list isn't exhasutive since Arabic have some diacritics and rarely used characters in Unicode
+        isArabic :: Char -> Bool
+        isArabic c = elem c ['ا', 'ب', 'ت', 'ة', 'ث', 'ج', 'ح', 'خ', 'د', 'ذ', 'ر', 'ز', 'س', 'ش', 'ص', 'ض', 'ط', 'ظ', 'ع', 'غ', 'ف', 'ق', 'ك', 'ل', 'م', 'ن', 'ه', 'ي', 'ء', 'آ', 'أ', 'إ', 'ؤ', 'و', 'ئ', 'ى']
+
+        -- TODO: Add all Arabic proclitics
+        isArabicProclitic :: Char -> Bool
+        isArabicProclitic c = elem c ['و', 'ف', 'ل', 'ب', 'ك']
+
+        isArabicProclitic2 :: Char -> Char -> Bool
+        isArabicProclitic2 c1 c2 = elem c1 ['ا', 'ل'] && elem c2 ['ل']
+
+        -- TODO: Add all Arabic proclitics
```

Comment:
I think this means to say enclitics. And would it be hard to add the remaining enclitics at this point, before merge?